pkg/controller/download_request_controller.go

```diff
@@ -0,0 +1,288 @@
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	// always call done on this item, since if it fails we'll add
+	// it back with rate-limiting below
+	defer c.queue.Done(key)
+
+	err := c.syncHandler(key.(string))
+	if err == nil {
+		// If you had no error, tell the queue to stop tracking history for your key. This will reset
+		// things like failure counts for per-item rate limiting.
+		c.queue.Forget(key)
+		return true
+	}
+
+	glog.Errorf("syncHandler error: %v", err)
+	// we had an error processing the item so add it back
+	// into the queue for re-processing with rate-limiting
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+func (c *downloadRequestController) processDownloadRequest(key string) error {
+	glog.V(4).Infof("processDownloadRequest for key %q", key)
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		glog.V(4).Infof("error splitting key %q: %v", key, err)
+		return err
+	}
+
+	downloadRequest, err := c.downloadRequestLister.DownloadRequests(ns).Get(name)
+	if apierrors.IsNotFound(err) {
+		glog.V(4).Infof("unable to find DownloadRequest %q", key)
+		return nil
+	}
+	if err != nil {
+		glog.V(4).Infof("error getting DownloadRequest %q: %v", key, err)
+		return err
+	}
+
+	switch downloadRequest.Status.Phase {
+	case "", v1.DownloadRequestPhaseNew:
+		return c.generatePreSignedURL(downloadRequest)
+	case v1.DownloadRequestPhaseProcessed:
+		return c.deleteIfExpired(downloadRequest)
+	}
+
+	return nil
+}
+
+const signedURLTTL = 10 * time.Minute
+
+func (c *downloadRequestController) generatePreSignedURL(downloadRequest *v1.DownloadRequest) error {
+	switch downloadRequest.Spec.Target.Kind {
+	case v1.DownloadTargetKindBackupLog:
+		update, err := cloneDownloadRequest(downloadRequest)
+		if err != nil {
+			return err
+		}
+
+		update.Status.DownloadURL, err = c.backupService.CreateBackupLogSignedURL(c.bucket, update.Spec.Target.Name, signedURLTTL)
+		if err != nil {
+			return err
+		}
+
+		update.Status.Phase = v1.DownloadRequestPhaseProcessed
+		update.Status.Expiration = metav1.NewTime(c.clock.Now().Add(signedURLTTL))
+
+		_, err = c.downloadRequestClient.DownloadRequests(update.Namespace).Update(update)
+		return err
+	}
+
+	return fmt.Errorf("unsupported download target kind %q", downloadRequest.Spec.Target.Kind)
+}
+
+func (c *downloadRequestController) deleteIfExpired(downloadRequest *v1.DownloadRequest) error {
+	glog.V(4).Infof("checking for expiration of %s/%s", downloadRequest.Namespace, downloadRequest.Name)
+	if downloadRequest.Status.Expiration.Time.Before(c.clock.Now()) {
+		glog.V(4).Infof("%s/%s has not expired", downloadRequest.Namespace, downloadRequest.Name)
+		return nil
+	}
+
+	glog.V(4).Infof("%s/%s has expired - deleting", downloadRequest.Namespace, downloadRequest.Name)
+	return c.downloadRequestClient.DownloadRequests(downloadRequest.Namespace).Delete(downloadRequest.Name, nil)
+}
+
+func (c *downloadRequestController) resync() {
```

Comment:
I did it this way because I intend for the `ark` client to manage the lifespan of a `DownloadRequest`:

1. client creates `DownloadRequest`
1. client waits for it to be `Processed`
1. controller processes it
1. client sees updated `DownloadRequest` and downloads the URL
1. client deletes the `DownloadRequest`

The resync is a failsafe that is meant to clean up any `DownloadRequest`s that the client failed to delete. I'm happy to put in a comment to this effect.

Does this make sense?